### PR TITLE
Add htmlParser, htmlSerializer, jsParser, jsSerializer options to middleware

### PR DESCRIPTION
**Pull request by usergenic**
Added support for custom parsing/pre-processing and post-processing/serializing functions to be provided to middleware to cover cases like babel parsing using syntax plugins.

### CODE REVIEW

**Comment by usergenic** [comment]
@justinfagnani Have a look at the README changes before the rest of the code.  Most of the code changes were some minor refactorings and renamings.  The more important bit here are the API changes-- specifically, addressing babel parsing by supporting a `jsParser` option that allows devs to do whatever they want to produce the AST.

File: src/support/string-utils.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+export const preserveSurroundingWhitespace =
+    (original: string, target: string): string => original.match(/^\s*/)![0] +
+    target.replace(/^\s*|\s*$/, '') + original.match(/\s*$/)![0];
```
Comment (aomarks) [inline src/support/string-utils.ts]: Need to add `/g` to the replace regexp or else this only matches leading whitespace.

File: README.md [hunk 2]
```diff
@@ -44,6 +45,50 @@ $ node dev-server.js
...
 Now you can serve up your web assets and Node package specifiers will be transformed on request.
 
+## Configuration
+
+`nodeResolve(options={})`
+
+### Options
+
+ - `root` the on-disk directory that maps to the served root URL, used to resolve module specifiers in filesystem.  In most cases this should match the root directory configured in your downstream static file server middleware.
+ - `logger` an alternative logger to use (`console` is the default).
```
[the remark this answers is not in the document]
Comment (usergenic) [reply README.md]: Super-updated the README with this description:

- `logger` an alternative logger to use (`console` is the default).  The logger will receive `error()` to record exceptions during parsing/transforming of JavaScript modules, `warn()` when a specifier is unresolvable, `info()` to report URLs with transformed content, `debug()` to report all Node module specifier resolutions.  All log messages are prefixed with `[koa-node-resolve]`.  To disable all logging, provide an empty object `{}`.

File: README.md [hunk 3]
```diff
@@ -44,6 +45,50 @@ $ node dev-server.js
...
+    import { removeFakeRootElements } = 'koa-node-resolve/lib/support/parse5-utils';
+
+    nodeResolve({ htmlParser: (source) => {
+      const ast = parse(source);
+      // Stop parse5 from creating <html>, <head> and <body> tags.
+      removeFakeRootElements(ast);
+      return ast;
+    });
+    ```
+ - `htmlSerializer` function to generate string from AST of `Parse5.DefaultTreeNode`.  The default implementation is equivalent to:
```
[the remark this answers is not in the document]
Comment (usergenic) [reply README.md]: I added details to the README but here's what I wrote after relocating fake element removal to the serialization phase:

- `htmlParser` function to convert HTML source to a `Parse5.DefaultTreeNode`.  The default implementation is equivalent to:
    ```js
    const { parse } = require('parse5');

    nodeResolve({ htmlParser: parse);
    ```

 - `htmlSerializer` function to generate string from AST of `Parse5.DefaultTreeNode`.  The default implementation is equivalent to:
    ```js
    const { serialize } = require('parse5');
    import { removeFakeRootElements } = 'koa-node-resolve/lib/support/parse5-utils';

    nodeResolve({ htmlSerializer: (ast) => {
      // Don't emit parse5's hallucenated <html>, <head> and <body> tags.
      removeFakeRootElements(ast);
      return serialize(ast); 
    });
    ```
    Maybe you don't *want* to remove the generated `<html>`, `<head>` and `<body>` tags-- you can provide an htmlSerializer function that doesn't do that then.

[the remark this answers is not in the document]
Comment (usergenic) [reply README.md]: Also for JS serialization:

 - `jsSerializer` function to generate output string from AST of `Babel.Node`.  The default implementation is equivalent to:
    ```js
    const serialize = require('@babel/generator');

    nodeResolve({ jsSerializer: (ast) => serialize(ast).code })
    ```
    The [@babel/generator documentation](https://babeljs.io/docs/en/babel-generator) shows many options available for tweaking output.  Maybe you want to generate source maps and embed them in the response; you could do that here.

File: src/support/string-utils.ts [hunk 4]
```diff
@@ -0,0 +1,3 @@
+export const preserveSurroundingWhitespace =
```
[the remark this answers is not in the document]
Comment (usergenic) [reply src/support/string-utils.ts]: Added!

File: src/support/string-utils.ts (hunk 1 above)
Comment (usergenic) [reply src/support/string-utils.ts]: I AM SO EMBARASSED.

Fixed!

File: README.md [hunk 5]
```diff
@@ -44,6 +45,50 @@ $ node dev-server.js
 
 Now you can serve up your web assets and Node package specifiers will be transformed on request.
 
+## Configuration
+
+`nodeResolve(options={})`
+
+### Options
+
+ - `root` the on-disk directory that maps to the served root URL, used to resolve module specifiers in filesystem.  In most cases this should match the root directory configured in your downstream static file server middleware.
```
[the remark this answers is not in the document]
Comment (usergenic) [reply README.md]: Fixd!

**Comment by usergenic** [comment]
ty!

**Pull request merged**